Test framework xml tests

mrvoorhe opened the pull request:
This PR contains new tests plus fixes for two tests that were failing

* Fix for all members of a type being preserved when a type is preserved with nothing.  This fixes the test UnusedTypeWithPreserveNothingHasMembersRemoved

* Fix for backing fields being preserved when a type is preserved with "methods".  If all of the properties were removed, then there's no reason to keep the backing fields

marek-safar commented:
build

marek-safar commented on linker/Mono.Linker.Steps/MarkStep.cs:
```diff
@@ -1150,14 +1150,20 @@ void ApplyPreserveMethods (MethodDefinition method)
 			foreach (FieldDefinition field in type.Fields) {
 				if (!includeStatic && field.IsStatic)
 					continue;
+				if (markBackingFieldsOnlyIfPropertyMarked && field.Name.EndsWith ("k__BackingField")) {
```
Should check for `">k__BackingField"`

mrvoorhe replied:
Done

marek-safar commented on linker/Mono.Linker.Steps/MarkStep.cs:
```diff
@@ -1150,14 +1150,20 @@ void ApplyPreserveMethods (MethodDefinition method)
 					continue;
+				if (markBackingFieldsOnlyIfPropertyMarked && field.Name.EndsWith ("k__BackingField")) {
+					var propertyName = field.Name.Substring (1, field.Name.IndexOf ('>') - 1);
+					var propertyDefinition = type.Properties.First (p => p.Name == propertyName);
```
This would throw if there isn't one

mrvoorhe replied:
If the backing field was there, is there a case where the property would not be there?

marek-safar replied:
Yes, compiler could reuse this name for something else or some other tool could generated only field with the name.

Also `field.Name.IndexOf ('>')` check fails with code like

```
interface I<T>
{
	int Foo { get; set; }
}

class X : I<int>
{
	int I<int>.Foo { get; set; }
}

```

marek-safar commented:
build

mrvoorhe commented:
Fixed up issues with the backing field logic in the MarkStep.  
* Don't throw if the property cannot be found
* Punting on handling cases where the property is an explicit implementation and the type has system types in the generic arguments.  The backing field will have a name of <IFoo<int>.Bar>k__BackingField while obtaining the property definition by string name requires the format IFoo<System.Int32>.Bar.  With some extra effort we could deal with this situation, but it's not worth the effort at the moment.

Added a new test called TypeWithPreserveFieldsHasBackingFieldsOfPropertiesRemoved to cover more complicated backing field situations.

Made dealing with backing fields in test cases easier.  Depending on the [KeptMember] attributes was getting out of control in some of the more complicated cases.  Writing new tests should be easy and getting the exact format of some of these complex signatures right is not easy.  Let the testing infrastructure figure out the names for us.

marek-safar commented:
whitelist

mrvoorhe commented:
build

mrvoorhe commented:
build

mrvoorhe commented:
@marek-safar I went ahead and implemented support for finding the PropertyDefinitions of backing fields such as `<IFoo<int>.Bar>k__BackingField`.  That was the easiest way to get the tests passing with both mcs & csc.

marek-safar commented on linker/Mono.Linker.Steps/MarkStep.cs:
```diff
@@ -1150,18 +1150,52 @@ void ApplyPreserveMethods (MethodDefinition method)
 			}
 		}
 
+		private PropertyDefinition SearchPropertiesForMatchingFieldDefinition (FieldDefinition field)
```
could be static

marek-safar commented on linker/Mono.Linker.Steps/MarkStep.cs:
```diff
@@ -1150,18 +1150,52 @@ void ApplyPreserveMethods (MethodDefinition method)
+					// Parsing the string and checking for these system types and converting the names would be ugly and error prone.
+					// Instead, let's search the properties and find the one that makes use of the current backing field
+					if (propertyDefinition == null) {
+						propertyDefinition = SearchPropertiesForMatchingFieldDefinition (field);
```
Perhaps it'd be better to maintain only one way of propertyDefinition look up and this one should cover both cases

mrvoorhe commented:
build

mrvoorhe commented:
@marek-safar  changes made
